### ISSUE TITLE
assert_eq! output is very unhelpful for floats

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#20870]
```
assert_eq!(1E-7, 0.0);
```

will output

> assertion failed: \`(left == right) && (right == left)\` (left: \`0f64\`, right: \`0f64\`)' …

Note that values displayed are pretty useless.

**Comment by kmcallister** [comment]
How about

```rust
trait DebugCompare<RHS> {
    pub describe_difference(&self, rhs: &RHS) -> Option<Cow<str>>;
}

macro_rules! assert_eq {
    ($left:expr, $right:expr) => {{
        match (&($left), &($right)) {
            (left_val, right_val) => if let Some(desc) = left_val.describe_difference(right_val) {
                panic!("assert_eq! failed: difference between {:?} and {:?} is {}",
                    *left_val, *right_val, desc);
            },
        }
    }}
}
```

Then `DebugCompare for f64` can give the numerical difference, `DebugCompare for Expr` can do an AST diff (???), etc.

**Comment by lifthrasiir** [comment]
With #24612 this prints the following, thus I believe this is fixed by now.

```
thread '<main>' panicked at 'assertion failed: `(left == right)` (left: `0.0000001`, right: `0`)', t.rs:1
```

**Comment by bstrie** [comment]
Closed by #24612.
